test/functional/options/defaults_spec.lua

```diff
@@ -328,3 +332,235 @@ describe('XDG-based defaults', function()
+    end)
+
+    it("doesn't expand $VARIABLES", function()
+      clear({env={
+        XDG_DATA_HOME='$VARIABLES',
+        VARIABLES='this-should-not-happen'
+      }})
+      eq(alter_path('$VARIABLES' .. appended_dir), funcs.stdpath('data'))
+    end)
+
+    it("doesn't expand ~/", function()
+      clear({env={
+        XDG_DATA_HOME='~/frobnitz'
+      }})
+      eq(alter_path('~/frobnitz' .. appended_dir), funcs.stdpath('data'))
+    end)
+  end)
+
+  describe('"cache"' , function ()
+    it('knows XDG_CACHE_HOME', function()
+      clear({env={
+        XDG_CACHE_HOME=alter_path('/home/docwhat/.cache')
+      }})
+      eq(alter_path('/home/docwhat/.cache/nvim'), funcs.stdpath('cache'))
+    end)
+
+    it('handles changes during runtime', function()
+      clear({env={
+        XDG_CACHE_HOME=alter_path('/home/original')
+      }})
+      eq(alter_path('/home/original/nvim'), funcs.stdpath('cache'))
+      nvim('command', 'let $XDG_CACHE_HOME="/home/new"')
+      eq(alter_path('/home/new/nvim'), funcs.stdpath('cache'))
+    end)
+
+    it("doesn't expand $VARIABLES", function()
+      clear({env={
+        XDG_CACHE_HOME='$VARIABLES',
+        VARIABLES='this-should-not-happen'
+      }})
+      eq(alter_path('$VARIABLES/nvim'), funcs.stdpath('cache'))
+    end)
+
+    it("doesn't expand ~/", function()
+      clear({env={
+        XDG_CACHE_HOME='~/frobnitz'
+      }})
+      eq(alter_path('~/frobnitz/nvim'), funcs.stdpath('cache'))
+    end)
+  end)
+
+  describe('"config_dirs"' , function ()
+    it('knows XDG_CONFIG_DIRS', function()
+      clear({env={
+        XDG_CONFIG_DIRS=alter_path('/home/docwhat/.config:/etc/config')
+      }})
+      eq({
+        alter_path('/home/docwhat/.config/nvim'),
+        alter_path('/etc/config/nvim'),
+      }, funcs.stdpath('config_dirs'))
+    end)
+
+    it('handles changes during runtime', function()
+      clear({env={
+        XDG_CONFIG_DIRS=alter_path('/home/original:/etc/config')
+      }})
+      eq({
+        alter_path('/home/original/nvim'),
+        alter_path('/etc/config/nvim')
+      }, funcs.stdpath('config_dirs'))
+      nvim('command', 'let $XDG_CONFIG_DIRS="/home/new:/etc/config"')
+      eq({
+        alter_path('/home/new/nvim'),
+        alter_path('/etc/config/nvim')
+      }, funcs.stdpath('config_dirs'))
+    end)
+
+    it("doesn't expand $VAR and $IBLES", function()
+      clear({env={
+        XDG_CONFIG_DIRS='$VAR:$IBLES',
+        VAR='this-should-not-happen',
+        IBLES='neither-should-this'
+      }})
+      eq({
+        alter_path('$VAR/nvim'),
+        alter_path('$IBLES/nvim'),
+      }, funcs.stdpath('config_dirs'))
+    end)
+
+    it("doesn't expand ~/", function()
+      clear({env={
+        XDG_CONFIG_DIRS='~/.oldconfig:~/.olderconfig'
+      }})
+      eq({
+        alter_path('~/.oldconfig/nvim'),
+        alter_path('~/.olderconfig/nvim'),
+      }, funcs.stdpath('config_dirs'))
+    end)
+  end)
+
+  describe('"data_dirs"' , function ()
```

Comment:
Thanks for the generator suggestion.  I implemented something similar, splitting the two tests into separate `it` statements.

I'm not sure how to test _colon-separated variable contains no directories_ since that triggers the default behavior from `stdpaths.c`.  Should I re-implement the `xdg_defaults` struct (with its `#ifdef WIN32` logic)? 

And regarding _colon-separated variable contains item larger then MAXPATHL_: What should I expect?

`get_xdg_home` and `concat_fnames_realloc` don't do any checks for `MAXPATHL` at the moment.

I also dislike hard-coding a value (e.g. 8192) since that can vary from OS to OS.  Is there a way to query vim or lua for this value?